File: 链表/25.K个一组反转链表.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        dummy = ListNode(0, head)
        groupPrev = dummy

        while True:
            Kth = self.getKth(groupPrev, k)
            if not Kth:
                break
            groupNext = Kth.next

            # 反转节点并获取反转后的首尾节点
            newHead, newTail = self.reverse(groupPrev.next, groupNext)
            # 重新链接反转后的链表部分
            groupPrev.next = newHead
            newTail.next = groupNext
            # 更新groupPrev到下一个节点的位置
            groupPrev = newTail

        return dummy.next

    def getKth(self, node: ListNode, k: int) -> ListNode:
        while node and k:
            node = node.next
            k -= 1
        return node

    def reverse(self, start: ListNode, end: ListNode) -> (ListNode, ListNode):
        prev, cur = end, start
        while cur != end:
            temp = cur.next
            cur.next = prev
            prev = cur
            cur = temp
        return prev, start

File: 链表/25.K个一组反转链表.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var reverseKGroup = function(head, k) {
    let dummy = new ListNode(0, head);
    let groupPrev = dummy;
    while (true) {
        let kth = getKth(groupPrev, k);
        if (!kth) break;

        // 反转k个链表
        let groupNext = kth.next;
        //! start: groupPrev.next | end: gtoupNext 
        let [newHead, newTail] = reverse(groupPrev.next, groupNext);
        // 连接
        groupPrev.next = newHead;
        newTail.next = groupNext;
        // 移动groupPrev为下一组准备
        groupPrev = newTail;
    }
    return dummy.next;
};

function getKth(node, k) {
    while (k > 0 && node) {
        k--;
        node = node.next;
    }
    return node;
}

//* reverse: 将cur指向prev
function reverse(start, end) {
    let prev = end;
    let cur = start;
    while (cur !== end) {  //! end不应该改变
        let temp = cur.next;
        cur.next = prev;
        prev = cur;
        cur = temp;
    }
    return [prev, start]
}
